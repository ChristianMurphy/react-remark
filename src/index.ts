import {
  FunctionComponent,
  Fragment,
  ReactElement,
  createElement,
  useState,
  useEffect,
  useCallback,
} from 'react';
import unified, { PluggableList } from 'unified';
import remarkParse, { RemarkParseOptions } from 'remark-parse';
import { Options as RemarkRehypeOptions } from 'mdast-util-to-hast';
import remarkToRehype from 'remark-rehype';
import rehypeReact, { Options as RehypeReactOptions } from 'rehype-react';

export interface UseRemarkOptions {
  remarkParseOptions?: Partial<RemarkParseOptions>;
  remarkToRehypeOptions?: RemarkRehypeOptions;
  rehypeReactOptions?: Partial<RehypeReactOptions<typeof createElement>>;
  remarkPlugins?: PluggableList;
  rehypePlugins?: PluggableList;
  onError?: (err: Error) => void;
}

export const useRemark = ({
  remarkParseOptions,
  remarkToRehypeOptions,
  rehypeReactOptions,
  remarkPlugins = [],
  rehypePlugins = [],
  onError = () => {},
}: UseRemarkOptions = {}): [ReactElement | null, (source: string) => void] => {
  const [reactContent, setReactContent] = useState<ReactElement | null>(null);

  const setMarkdownSource = useCallback((source: string) => {
    unified()
      .use(remarkParse, remarkParseOptions)
      .use(remarkPlugins)
      .use(remarkToRehype, remarkToRehypeOptions)
      .use(rehypePlugins)
      .use(rehypeReact, { createElement, Fragment, ...rehypeReactOptions })
      .process(source)
      .then((vfile) => setReactContent(vfile.result as ReactElement))
      .catch(onError);
  }, []);

  return [reactContent, setMarkdownSource];
};

export interface RemarkProps extends UseRemarkOptions {
  children: string;
}

export const Remark: FunctionComponent<RemarkProps> = ({
  children,
  ...useRemarkOptions
}: RemarkProps) => {
  const [reactContent, setMarkdownSource] = useRemark(useRemarkOptions);

  useEffect(() => {
    setMarkdownSource(children);
  }, [children, setMarkdownSource]);

  return reactContent;
};
