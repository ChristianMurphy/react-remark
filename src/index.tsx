import React, {
  FunctionComponent,
  ReactElement,
  useState,
  useEffect,
  useCallback,
} from 'react';
import unified, { PluggableList } from 'unified';
import remarkParse, { RemarkParseOptions } from 'remark-parse';
// @ts-ignore
import remarkToRehype from 'remark-rehype';
// @ts-ignore
import rehypeReact from 'rehype-react';

export interface UseRemarkOptions {
  remarkParseOptions?: Partial<RemarkParseOptions>;
  // TODO: Pending https://github.com/remarkjs/remark-rehype/pull/13
  remarkToRehypeOptions?: {};
  // TODO: Pending
  rehypeReactOptions?: {};
  remarkPlugins?: PluggableList;
  rehypePlugins?: PluggableList;
  onError?: (err: Error) => void;
}

export const useRemark = ({
  remarkParseOptions,
  remarkToRehypeOptions,
  rehypeReactOptions = { createElement: React.createElement },
  remarkPlugins = [],
  rehypePlugins = [],
  onError = () => {},
}: UseRemarkOptions = {}): [ReactElement | null, (source: string) => void] => {
  const [reactContent, setReactContent] = useState<ReactElement | null>(null);

  const setMarkdownSource = useCallback((source: string) => {
    unified()
      .use(remarkParse, remarkParseOptions)
      .use(remarkPlugins)
      .use(remarkToRehype, remarkToRehypeOptions)
      .use(rehypePlugins)
      .use(rehypeReact, rehypeReactOptions)
      .process(source)
      // @ts-ignore pending https://github.com/vfile/vfile/pull/53
      .then(vfile => setReactContent(vfile.result))
      .catch(onError);
  }, []);

  return [reactContent, setMarkdownSource];
};

export interface RemarkProps extends UseRemarkOptions {
  children: string;
}

export const Remark: FunctionComponent<RemarkProps> = ({
  children,
  ...useRemarkOptions
}: RemarkProps) => {
  const [reactContent, setMarkdownSource] = useMarkdown(useRemarkOptions);

  useEffect(() => {
    setMarkdownSource(children);
  }, [children, setMarkdownSource]);

  return reactContent;
};
