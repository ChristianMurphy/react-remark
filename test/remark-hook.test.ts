import { createElement } from 'react';
import type { ComponentPropsWithoutNode } from 'rehype-react';
import { renderHook, act } from '@testing-library/react-hooks';
import '@testing-library/jest-dom/extend-expect';
import remarkGfm from 'remark-gfm';
import rehypeRaw from 'rehype-raw';
import rehypeSanitize from 'rehype-sanitize';
import remarkMath from 'remark-math';
import rehypeKatex from 'rehype-katex';
import { useRemark } from '../src';

describe('useRemark', () => {
  it('should render content', async () => {
    const { result, waitForNextUpdate } = renderHook(() => useRemark());
    act(() => {
      result.current[1]('# header');
    });
    await waitForNextUpdate();
    expect(result.current[0]).toMatchSnapshot();
  });

  it('should support gfm through remark plugins', async () => {
    const { result, waitForNextUpdate } = renderHook(() =>
      useRemark({ remarkPlugins: [remarkGfm] })
    );
    act(() => {
      result.current[1]('https://example.com');
    });
    await waitForNextUpdate();
    expect(result.current[0]).toMatchSnapshot();
  });

  it('should support html through rehype plugins', async () => {
    const { result, waitForNextUpdate } = renderHook(() =>
      useRemark({
        remarkToRehypeOptions: { allowDangerousHtml: true },
        rehypePlugins: [rehypeRaw, rehypeSanitize],
      })
    );
    act(() => {
      result.current[1]('<span>example</span>');
    });
    await waitForNextUpdate();
    expect(result.current[0]).toMatchSnapshot();
  });

  it('should support math through remark and rehype plugins', async () => {
    const { result, waitForNextUpdate } = renderHook(() =>
      useRemark({
        remarkPlugins: [remarkMath],
        rehypePlugins: [rehypeKatex],
      })
    );
    act(() => {
      result.current[1](
        'Lift($L$) can be determined by Lift Coefficient ($C_L$) like the following equation.'
      );
    });
    await waitForNextUpdate();
    expect(result.current[0]).toMatchSnapshot();
  });

  it('should support custom element renderer', async () => {
    const { result, waitForNextUpdate } = renderHook(() =>
      useRemark({
        rehypeReactOptions: {
          components: {
            h1: (props: ComponentPropsWithoutNode) =>
              createElement('h2', props),
          },
        },
      })
    );
    act(() => {
      result.current[1]('# heading');
    });
    await waitForNextUpdate();
    expect(result.current[0]).toMatchSnapshot();
  });
});
