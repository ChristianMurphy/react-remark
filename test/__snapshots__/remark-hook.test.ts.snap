// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRemark should render content 1`] = `
<React.Fragment>
  <h1>
    header
  </h1>
</React.Fragment>
`;

exports[`useRemark should support custom element renderer 1`] = `
<React.Fragment>
  <h1>
    heading
  </h1>
</React.Fragment>
`;

exports[`useRemark should support gfm through remark plugins 1`] = `
<React.Fragment>
  <p>
    <a
      href="https://example.com"
    >
      https://example.com
    </a>
  </p>
</React.Fragment>
`;

exports[`useRemark should support html through rehype plugins 1`] = `
<React.Fragment>
  <p>
    <span>
      example
    </span>
  </p>
</React.Fragment>
`;

exports[`useRemark should support math through remark and rehype plugins 1`] = `
<React.Fragment>
  <p>
    Lift(
    <span
      className="math math-inline"
    >
      <span
        className="katex"
      >
        <span
          className="katex-mathml"
        >
          <math
            xmlns="http://www.w3.org/1998/Math/MathML"
          >
            <semantics>
              <mrow>
                <mi>
                  L
                </mi>
              </mrow>
              <annotation
                encoding="application/x-tex"
              >
                L
              </annotation>
            </semantics>
          </math>
        </span>
        <span
          aria-hidden="true"
          className="katex-html"
        >
          <span
            className="base"
          >
            <span
              className="strut"
              style={
                Object {
                  "height": "0.68333em",
                  "verticalAlign": "0em",
                }
              }
            />
            <span
              className="mord mathnormal"
            >
              L
            </span>
          </span>
        </span>
      </span>
    </span>
    ) can be determined by Lift Coefficient (
    <span
      className="math math-inline"
    >
      <span
        className="katex"
      >
        <span
          className="katex-mathml"
        >
          <math
            xmlns="http://www.w3.org/1998/Math/MathML"
          >
            <semantics>
              <mrow>
                <msub>
                  <mi>
                    C
                  </mi>
                  <mi>
                    L
                  </mi>
                </msub>
              </mrow>
              <annotation
                encoding="application/x-tex"
              >
                C_L
              </annotation>
            </semantics>
          </math>
        </span>
        <span
          aria-hidden="true"
          className="katex-html"
        >
          <span
            className="base"
          >
            <span
              className="strut"
              style={
                Object {
                  "height": "0.83333em",
                  "verticalAlign": "-0.15em",
                }
              }
            />
            <span
              className="mord"
            >
              <span
                className="mord mathnormal"
                style={
                  Object {
                    "marginRight": "0.07153em",
                  }
                }
              >
                C
              </span>
              <span
                className="msupsub"
              >
                <span
                  className="vlist-t vlist-t2"
                >
                  <span
                    className="vlist-r"
                  >
                    <span
                      className="vlist"
                      style={
                        Object {
                          "height": "0.32833099999999993em",
                        }
                      }
                    >
                      <span
                        style={
                          Object {
                            "marginLeft": "-0.07153em",
                            "marginRight": "0.05em",
                            "top": "-2.5500000000000003em",
                          }
                        }
                      >
                        <span
                          className="pstrut"
                          style={
                            Object {
                              "height": "2.7em",
                            }
                          }
                        />
                        <span
                          className="sizing reset-size6 size3 mtight"
                        >
                          <span
                            className="mord mathnormal mtight"
                          >
                            L
                          </span>
                        </span>
                      </span>
                    </span>
                    <span
                      className="vlist-s"
                    >
                      â€‹
                    </span>
                  </span>
                  <span
                    className="vlist-r"
                  >
                    <span
                      className="vlist"
                      style={
                        Object {
                          "height": "0.15em",
                        }
                      }
                    >
                      <span />
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </span>
        </span>
      </span>
    </span>
    ) like the following equation.
  </p>
</React.Fragment>
`;
